### PR TITLE
Add a PartitionRouting partition-id-to-node mapping holder and background refresher

### PR DESCRIPTION
**Pull request by pcholakov**
New approach as an alternative to https://github.com/restatedev/restate/pull/2149, which exposes a dedicated `PartitionRouting` lookup interface. Its implementation is initially backed by the Metadata Store but it is not expected to stay that way for long.

### CODE REVIEW

File: crates/core/src/metadata/manager.rs
Comment (pcholakov) [inline crates/core/src/metadata/manager.rs:49]: Noticed this is unused, so removed it.

File: crates/core/src/routing_info/mod.rs
Comment (pcholakov) [inline crates/core/src/routing_info/mod.rs:123]: Wasn't sure what to pick for this. It seems reasonable to start with; it might be a nice optimisation to add a `get_if_newer_version` to the metadata store to make it cheaper to poll. Eventually we should gossip/push this info to nodes without needing to poll all the time.

File: crates/core/src/routing_info/mod.rs
Comment (pcholakov) [inline crates/core/src/routing_info/mod.rs:129]: Is this a reasonable way to implement this? I based it off the `MetadataManager`.

File: crates/core/src/routing_info/mod.rs
Comment (pcholakov) [inline crates/core/src/routing_info/mod.rs:159]: I use `spawn_unmanaged` to get a task handle which allows me to check if the previous one is finished.

File: crates/core/src/routing_info/mod.rs
Comment (pcholakov) [inline crates/core/src/routing_info/mod.rs:46]: I'm not in love with this name but hate everything else I've come up with more. Any suggestions? `PartitionNodeRoutes` is a more verbose version I also like.

File: crates/core/src/routing_info/mod.rs
Comment (tillrohrmann) [inline crates/core/src/routing_info/mod.rs:39]: Why passing in the `TargetVersion`? It seems unused.

File: crates/core/src/routing_info/mod.rs
Comment (tillrohrmann) [inline crates/core/src/routing_info/mod.rs:65]: How will a user of `PartitionRouting` use it if they encounter invalid routing information?

File: crates/core/src/routing_info/mod.rs
Comment (tillrohrmann) [inline crates/core/src/routing_info/mod.rs:118]: Copy paste error?

File: crates/core/src/routing_info/mod.rs
Comment (tillrohrmann) [reply crates/core/src/routing_info/mod.rs:129]: Yes, looks good to me :-)

File: crates/core/src/routing_info/mod.rs
Comment (tillrohrmann) [reply crates/core/src/routing_info/mod.rs:123]: Wondering whether the periodic update is really required if users of this struct refresh explicitly on outdated leader information.

File: crates/core/src/routing_info/mod.rs
Comment (tillrohrmann) [inline crates/core/src/routing_info/mod.rs:201]: What if there is no scheduler yet that has created a `SchedulingPlan`?

File: crates/core/src/routing_info/mod.rs
Comment (tillrohrmann) [inline crates/core/src/routing_info/mod.rs:196]: What is this parameter used for? It looks that all call-sites pass in `None`.

File: crates/core/src/routing_info/mod.rs
Comment (tillrohrmann) [inline crates/core/src/routing_info/mod.rs:208]: Optimization: If we remember the last `SchedulingPlan` version then we can check and only allocate a new HashMap and extract the information if the version is newer.

File: crates/types/src/cluster_controller.rs
Comment (tillrohrmann) [inline crates/types/src/cluster_controller.rs:40]: Why do you need to make the fields public?

File: crates/core/src/routing_info/mod.rs
Comment (tillrohrmann) [reply crates/core/src/routing_info/mod.rs:159]: Good idea 👍

**Review by tillrohrmann — COMMENTED** [review]
Thanks for creating this PR @pcholakov. I think this goes exactly in the direction I had in mind for the simple solution to unblock Igal's and Francesco's work. I left a few comments and suggestions we could do before merging.

File: crates/core/src/routing_info/mod.rs
Comment (pcholakov) [reply crates/core/src/routing_info/mod.rs:65]: Just call it? :-) It's super dumb right now, just a fire and forget to trigger the refresh. We don't yet have anything errors that would signify that one is working with outdated routing info but I imagine that instead of `TargetVersion::Latest`, the caller might specify something they got from a peer, such as monotonic a partition mapping version/epoch number. But the expectation is that whoever calls this already encountered an unrecoverable error, and calling this is just a nice to have signal. Let me see if I can work this into a doc comment.

File: crates/core/src/routing_info/mod.rs
Comment (pcholakov) [reply crates/core/src/routing_info/mod.rs:208]: Ack, already on my list.

File: crates/types/src/cluster_controller.rs
Comment (pcholakov) [reply crates/types/src/cluster_controller.rs:40]: `sync_routing_information` iterates over the items in `partitions` directly - seemed like I might as well make `version` private at the same time. On reflection, it seems like a more Rust-y approach would be to either implement `IntoIterator` or expose an `iter()` method from `SchedulingPlan` - which it turns out, you already thought about :-) Totally missed the existing `iter()`, will revert!

File: crates/core/src/routing_info/mod.rs
Comment (pcholakov) [reply crates/core/src/routing_info/mod.rs:123]: I don't believe we have any explicit mechanism today to provide outdated feedback; I exposed the refresh method just in case, but let's keep this for now please. Happy to reduce the refresh interval to something like 3-5x the heartbeat interval initially.

File: crates/core/src/routing_info/mod.rs
Comment (pcholakov) [reply crates/core/src/routing_info/mod.rs:201]: I figure we'll just fail some requests until we have one - if a caller is trying to look up routing info for a partition, and we don't yet know how to address any of its nodes, it should return an appropriate error. E.g. the ingress layer might just return a 500 server-internal retryable error indicating to the upstream caller that the condition should be transient.

Comment (pcholakov) [reply crates/core/src/routing_info/mod.rs:201]: I will add a debug-level log message about this at least.

File: crates/core/src/routing_info/mod.rs
Comment (tillrohrmann) [reply crates/core/src/routing_info/mod.rs]: What do I do after I called this method? Right now, I cannot wait until the update has been processed. So I can only continue and hope that by the time I retry, the information has been updated. If not, then I'll talk to the outdated leader again.

Comment (tillrohrmann) [reply crates/core/src/routing_info/mod.rs]: I was imagining the caller having a number of retries. Whenever it receives a "wrong node" response indicating that it is talking to a wrong node, it would like to refresh the routing information hoping that it was operating against wrong information. Once new information is obtained, it could continue hoping that this time the information is up to date.

If the caller stopped retrying, then only notifying the `RoutingInformation` that the data is outdated is probably enough.

**Comment by tillrohrmann** [comment]
> In this PR I can't find where I would be able to acces this `PartitionRouting`? Is it something I need to instantiate inside the ingress? Or will I be able to get a handle to that with a thread local or alike?

I guess it will be passed into the `Ingress`.

**Comment by pcholakov** [comment]
@slinkydeveloper see how I've wired it up in this PR: https://github.com/restatedev/restate/pull/2172

File: crates/core/src/routing_info/mod.rs
Comment (pcholakov) [reply crates/core/src/routing_info/mod.rs]: @tillrohrmann I think this is the main outstanding concern with this PR - would you like to see some specific improvements before we can merge this? I thought about this a fair amount yesterday and don't believe we can meaningfully improve on "just back off and retry" at the moment, though we'll definitely be able to do better in the future once we have better view of cluster node status.

Comment (tillrohrmann) [reply crates/core/src/routing_info/mod.rs]: We can take this as a follow-up. One idea could be to have a watch that gets updated after the refresh task completes. A caller could then await this update, for example.

Comment (pcholakov) [reply crates/core/src/routing_info/mod.rs]: Sounds good - I'd love to do that follow-up!

File: crates/core/src/routing_info/mod.rs
Comment (tillrohrmann) [reply crates/core/src/routing_info/mod.rs:124]: Let's see whether the load on the metadata store will be too high. Something to keep an eye out for.

File: crates/core/src/routing_info/mod.rs
Comment (tillrohrmann) [inline crates/core/src/routing_info/mod.rs:194]: Maybe we also log on debug if an error occurred here. That way things might become easier to debug if the routing does not update.

File: crates/core/src/routing_info/mod.rs
Comment (tillrohrmann) [inline crates/core/src/routing_info/mod.rs:222]: Nice :-)

File: crates/core/src/routing_info/mod.rs
Comment (tillrohrmann) [inline crates/core/src/routing_info/mod.rs:69]: Nit: Since the `Command` no longer carries any specific value, one could use a watch to signal the refresh request. That way, the `request_refresh` call could become synchronous and users of this method wouldn't block on the availability of send permits.

**Review by tillrohrmann — APPROVED** [review]
Thanks for implementing the `PartitionRouting` struct @pcholakov. The changes look really good. +1 for merging.

File: crates/core/src/routing_info/mod.rs
Comment (pcholakov) [reply crates/core/src/routing_info/mod.rs:69]: That's a neat use of watch! I do think the sync command will grow to carry some information again in the future - perhaps a node id, partition id, or some other context from the caller - as we learn more about the conditions that should trigger a refresh. So, I'd rather keep in as is for the time being, even though it's a totally valid nit!

File: crates/core/src/routing_info/mod.rs
Comment (pcholakov) [reply crates/core/src/routing_info/mod.rs:124]: Yeah! Not entirely certain what signals to monitor that will tell us that this is a problem, before it becomes an actual problem, if that makes sense. Probably some metrics around metadata store service?

I'm pretty keen on the `get_if_version_newer` optimisation - I think basically any backing store we would choose for metadata should easily support this feature, and worst case, just fall back on always doing a full read if it can't. At least saves some bytes over the wire.

**Pull request merged**
